clippy_lints/src/manual_let_else.rs

```diff
@@ -0,0 +1,269 @@
+use clippy_utils::diagnostics::span_lint;
+use clippy_utils::higher::IfLetOrMatch;
+use clippy_utils::ty::is_type_diagnostic_item;
+use clippy_utils::visitors::expr_visitor_no_bodies;
+use clippy_utils::{meets_msrv, msrvs, peel_blocks};
+use if_chain::if_chain;
+use rustc_data_structures::stable_set::FxHashSet;
+use rustc_hir::intravisit::Visitor;
+use rustc_hir::{Expr, ExprKind, MatchSource, Pat, PatKind, QPath, Stmt, StmtKind};
+use rustc_lint::{LateContext, LateLintPass, LintContext};
+use rustc_middle::lint::in_external_macro;
+use rustc_semver::RustcVersion;
+use rustc_session::{declare_tool_lint, impl_lint_pass};
+use rustc_span::symbol::sym;
+use rustc_span::Span;
+
+declare_clippy_lint! {
+    /// ### What it does
+    ///
+    /// Warn of cases where `let else` could be used
+    ///
+    /// ### Why is this bad?
+    ///
+    /// `let else` is more compact.
+    ///
+    /// ### Example
+    ///
+    /// ```rust
+    /// # let w = Some(0);
+    /// let v = if let Some(v) = w { v } else { return };
+    /// ```
+    ///
+    /// Could be written:
+    ///
+    /// ```rust
+    /// # #![feature(let_else)]
+    /// # fn main () {
+    /// # let w = Some(0);
+    /// let Some(v) = w else { return };
+    /// # }
+    /// ```
+    #[clippy::version = "1.60.0"]
+    pub MANUAL_LET_ELSE,
+    restriction,
```

Comment:
> I think it can reasonably be a style lint.

I would be more careful, as there is no rustfmt support for it yet.

> Matching `let x = match` would lead to false positives

Note that I have severely restricted the lint for `match`.

Reply:
I've been making the assumption that this PR won't get merged until we have full support for let-else. Given that assumption, I think it can be a style lint.

Reply:
@joshtriplett good, we agree then :)